pkg/generate/git/git.go

```diff
@@ -5,35 +5,64 @@ import (
 	"io"
 	"net/url"
 	"path"
+	"regexp"
+	"runtime"
 	"strings"
-
-	s2igit "github.com/openshift/source-to-image/pkg/scm/git"
-	s2iutil "github.com/openshift/source-to-image/pkg/util"
 )
 
-// ParseRepository parses a string that may be in the Git format (git@) or URL format
-// and extracts the appropriate value. Any fragment on the URL is preserved.
+var gitURLRegexp = regexp.MustCompile("(?i)^[a-z][-a-z0-9+.]*://") // matches scheme:// according to RFC3986
+var dosDriveRegexp = regexp.MustCompile("(?i)^[a-z]:")
+var sshRegexp = regexp.MustCompile("^" +
+	"(?:([^@/]*)@)?" + // user@ (optional)
+	"([^/]*):" + //            host:
+	"([^#]*)" + //                  path
+	"(?:#(.*))?" + //                   #fragment (optional)
+	"$")
+var pathRegexp = regexp.MustCompile("^" +
+	"([^#]*)" + //        path
+	"(?:#(.*))?" + //         #fragment (optional)
+	"$")
+
+// ParseRepository parses a git repository spec.
+//
+// Git clone accepts repository specs in three broad formats:
+// 1) URL including '://', e.g. https://github.com/openshift/origin,
+//    file:///foo/bar, etc.
+// 2) SSH spec, including a ':' with no preceding '/', but not of the form C:...
+//    on Windows, e.g. git@github.com:openshift/origin, etc.
+// 3) OS-specific relative or absolute local file path, e.g. foo/bar,
+//    C:\foo\bar, etc.
+//
+// We extend all of the above specs to additionally accept an optional #fragment
+// appended to a given spec, which is given to specify a git reference.
 //
 // Protocols returned:
 // - http, https
 // - file
 // - git
 // - ssh
 func ParseRepository(s string) (*url.URL, error) {
-	uri, err := url.Parse(s)
-	if err != nil {
-		return nil, err
+	if gitURLRegexp.MatchString(s) {
+		return url.Parse(s)
 	}
-
-	// There are some shortcomings with url.Parse when it comes to GIT, namely wrt
-	// the GIT local/file and ssh protocols - it does not handle implied schema (i.e. no <proto>:// prefix)well;
-	// We handle those caveats here
-	err = s2igit.New(s2iutil.NewFileSystem()).MungeNoProtocolURL(s, uri)
-	if err != nil {
-		return nil, err
+	if m := sshRegexp.FindStringSubmatch(s); m != nil && runtime.GOOS != "windows" || dosDriveRegexp.MatchString(s) {
+		u := &url.URL{
+			Scheme:   "ssh",
+			Host:     m[2],
+			Path:     m[3],
+			Fragment: m[4],
+		}
+		if m[1] != "" {
+			u.User = url.User(m[1])
+		}
+		return u, nil
```

Comment:
100% agreed.  Having a single code path shared by both scenarios that addresses the multitude of pain points wrt URL parsing in golang should be the ultimate goal, and what I wanted (but may not have entirely succeeded) to convey.  I'll defer on whether it is done in one fell swoop or over a series of changes.